src/test/rbd_mirror/test_mock_ImageSyncThrottler.cc

```diff
@@ -0,0 +1,399 @@
+
+    mock_sync_throttler = new MockImageSyncThrottler();
+
+    m_mock_local_image_ctx = new librbd::MockImageCtx(*m_local_image_ctx);
+    m_mock_remote_image_ctx = new librbd::MockImageCtx(*m_remote_image_ctx);
+    m_mock_journaler = new journal::MockJournaler();
+  }
+
+  virtual void TearDown() {
+    MockImageSync::instances.clear();
+    delete mock_sync_throttler;
+    delete m_mock_local_image_ctx;
+    delete m_mock_remote_image_ctx;
+    delete m_mock_journaler;
+    TestMockFixture::TearDown();
+  }
+
+  void start_sync(const std::string& image_id, Context *ctx) {
+    m_mock_local_image_ctx->id = image_id;
+    mock_sync_throttler->start_sync(m_mock_local_image_ctx,
+                                    m_mock_remote_image_ctx,
+                                    m_threads->timer,
+                                    &m_threads->timer_lock,
+                                    "mirror_uuid",
+                                    m_mock_journaler,
+                                    &m_client_meta,
+                                    m_threads->work_queue,
+                                    ctx);
+  }
+
+  void cancel(const std::string& mirror_uuid, MockImageSync *sync,
+              bool running=true) {
+    if (running) {
+      EXPECT_CALL(*sync, cancel())
+        .WillOnce(Invoke([sync]() {
+              sync->finish(-ECANCELED);
+            }));
+    } else {
+      EXPECT_CALL(*sync, cancel()).Times(0);
+    }
+    mock_sync_throttler->cancel_sync(mirror_uuid);
+  }
+
+  librbd::ImageCtx *m_remote_image_ctx;
+  librbd::ImageCtx *m_local_image_ctx;
+  librbd::MockImageCtx *m_mock_local_image_ctx;
+  librbd::MockImageCtx *m_mock_remote_image_ctx;
+  journal::MockJournaler *m_mock_journaler;
+  librbd::journal::MirrorPeerClientMeta m_client_meta;
+  MockImageSyncThrottler *mock_sync_throttler;
+};
+
+TEST_F(TestMockImageSyncThrottler, Single_Sync) {
+  C_SaferCond ctx;
+  start_sync("image_id", &ctx);
+
+  ASSERT_EQ(1u, MockImageSync::instances.size());
+  MockImageSync *sync = MockImageSync::instances[0];
+  ASSERT_EQ(true, sync->syncing);
+  sync->finish(0);
+  ASSERT_EQ(0, ctx.wait());
+}
+
+TEST_F(TestMockImageSyncThrottler, Multiple_Syncs) {
+  mock_sync_throttler->set_max_concurrent_syncs(2);
+
+  C_SaferCond ctx1;
+  start_sync("image_id_1", &ctx1);
+  C_SaferCond ctx2;
+  start_sync("image_id_2", &ctx2);
+  C_SaferCond ctx3;
+  start_sync("image_id_3", &ctx3);
+  C_SaferCond ctx4;
+  start_sync("image_id_4", &ctx4);
+
+  ASSERT_EQ(4u, MockImageSync::instances.size());
+
+  MockImageSync *sync1 = MockImageSync::instances[0];
+  ASSERT_TRUE(sync1->syncing);
+
+  MockImageSync *sync2 = MockImageSync::instances[1];
+  ASSERT_TRUE(sync2->syncing);
+
+  MockImageSync *sync3 = MockImageSync::instances[2];
+  ASSERT_FALSE(sync3->syncing);
+
+  MockImageSync *sync4 = MockImageSync::instances[3];
+  ASSERT_FALSE(sync4->syncing);
+
+  sync1->finish(0);
+  ASSERT_EQ(0, ctx1.wait());
+
+  ASSERT_TRUE(sync3->syncing);
```

Comment:
Definitely a race condition here unless you swap the order of starting the next sync and invoking the on_finish of the finished sync (from comment below)